### PR TITLE
feat: Add bar chart version validators

### LINKED ISSUE #25: Server version chart
Per @crypticrabbit ( https://github.com/XRPLF/xrpl-dev-portal/issues/1161 )

> I think it would be useful (for things like amendment voting decisions) to have a chart showing what proportion of the network is running a given version of rippled on the livenet.xrpl.org/network page.
>
> Something similar to the chart in the top right of this page: https://xrpcharts.ripple.com/#/topology
>
> While the chart on the ripple.com page appears accurate, I think there are still multiple reasons (convenience, ensuring data accuracy, independent data, etc.) that having the chart on xrpl.org would be beneficial.
>
> Cheers!

### PR DESCRIPTION
**Pull request by pdp2121**
## High Level Overview of Change

Add bar chart showing versions of nodes on the network. Closes #25.

### Context of Change

When deciding whether to vote “Yes” or “No” on an amendment, it matters what version most servers on the network are running. If validators activate an amendment, people who don't have software understanding the amendment will be "amendment blocked".

This affects people's code and their ability to interact with the XRP Ledger.

XRPCharts used to have a graph which showed which versions everyone in the network was running: [XRP Charts](https://xrpcharts.ripple.com/#/topology) (GitHub repo: [GitHub - ripple/ripplecharts-frontend: RippleCharts front-end using Angular.js and D3](https://github.com/ripple/ripplecharts-frontend) ). It is now deprecated and out-of-date.

This adds something like that to the actively maintained explorer (this project) in order to let folks know when it is safe to vote “Yes” on upcoming amendments because enough of the ecosystem has upgraded to the necessary version of `rippled`.

The changes would also include refactors to use react hooks and typescript.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### TypeScript/Hooks Update

- [x] Updated files to React Hooks
- [x] Updated files to TypeScript

## Before / After

###After
Screen Shot 2022-09-08 at 11 06 09 AM (2)

## Test Plan

Create color code for the bar chart to indicate stable version upgrade progress.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -139,7 +151,7 @@
     "lint:js": "node_modules/.bin/eslint --ext=js --ext=jsx --ext=ts --ext=tsx --color .",
     "precommit": "lint-staged",
     "prod-server": "node server|bunyan",
-    "start": "DISABLE_NEW_JSX_TRANSFORM=true run-p dev-server dev-client",
+    "start": "DISABLE_NEW_JSX_TRANSFORM=true PORT=3006 run-p dev-server dev-client",
```
Comment (ckniffen) [inline package.json:154]: Remove this line.  I think you added this for some local dev.

File: src/containers/Network/BarChartVersion.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+const BarChartVersion = (props: Props) => {
+  const { data } = props
+  const { t } = useTranslation()
+  const [posData, setposData] = useState<BarCoordinates>({ x: 0, y: 0 })
+  return (
+    <div className="barchart">
+      <BarChart
+        data={data}
+        width={1075}
```
Comment (ckniffen) [inline src/containers/Network/BarChartVersion.tsx:54]: Is a width required?

File: src/containers/Network/NetworkTab.tsx [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import React from 'react'
+import Tabs from '../shared/components/Tabs'
+
+interface Props {
+  selected: string
```
Comment (ckniffen) [inline src/containers/Network/NetworkTab.tsx:5]: Derive `path` via  `const { path = '/' } = useRouteMatch()` instead of a prop.

**Comment by ckniffen** [comment]
Please remove `yarn.lock`.

File: src/containers/Network/index.js [hunk 4]
```diff
@@ -38,7 +39,14 @@ class Network extends Component {
     }
     // strips :url from the front and the tab info from the end
     const base = path.split('/:')[0]
-    return tab === 'nodes' ? <Nodes path={base} /> : <Validators path={base} />
+    // eslint-disable-next-line no-nested-ternary
+    return tab === 'nodes' ? (
```
Comment (ckniffen) [inline src/containers/Network/index.js:43]: Nodes should be the default display of the page.

File: src/containers/Network/css/barchart.scss [hunk 5]
```diff
@@ -0,0 +1,13 @@
+p {
```
Comment (ckniffen) [inline src/containers/Network/css/barchart.scss:1]: This will change all `<p>` tags globally.  You should make `p` and `.yLabel` styles use specific selectors.

File: src/containers/Network/upgradeStatus.tsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+          {vList && (
+            <BarChartVersion data={aggregateData(Object.values(vList))} />
+          )}
+        </div>
+      </div>
+    </div>
+  )
+}
+
+export default UpgradeStatus
```
Comment (ckniffen) [inline src/containers/Network/upgradeStatus.tsx:129]: Use named exports.

File: src/containers/Network/BarChartVersion.tsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+          dx={-7}
+          height={90}
+          tickLine={false}
+          stroke="#9BA2B0"
+        />
+        <YAxis
+          className="yAxis"
+          tickLine={false}
+          tickFormatter={(tick) => `${tick}%`}
+          stroke="#9BA2B0"
```
Comment (ckniffen) [inline src/containers/Network/BarChartVersion.tsx:71]: Use variables for colors.

**Comment by ckniffen** [comment]
How does this look on mobile?

**Comment by pdp2121** [comment]
> How does this look on mobile?

Fixed tick marks hidden for mobile screen

File: package.json [hunk 8]
```diff
@@ -5,13 +5,16 @@
   "dependencies": {
     "@fontsource/roboto": "^4.5.7",
     "@google-cloud/bigquery": "^5.5.0",
+    "@types/d3-hexbin": "^0.2.3",
```
Comment (mvadari) [inline package.json:8]: types should be dev dependencies

File: src/containers/Network/NetworkTab.tsx [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import React from 'react'
```
Comment (mvadari) [inline src/containers/Network/NetworkTab.tsx:1]: The refactors to TypeScript/Hooks should be a separate PR from the new feature, to make reviewing easier

Comment (ckniffen) [reply src/containers/Network/NetworkTab.tsx:1]: This was done at the same time because there was a lot of code that was shared with the validator list page.

Comment (ckniffen) [reply src/containers/Network/NetworkTab.tsx:1]: This is also a new component that was created to prevent the same code from being copied in 3 places.  Was two before the new tab.

File: src/containers/Network/NetworkTab.tsx [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+import { useRouteMatch } from 'react-router'
+import Tabs from '../shared/components/Tabs'
+
+interface Props {
+  selected: string
+}
+
+const NetworkTabs = (props: Props) => {
```
Comment (ckniffen) [inline src/containers/Network/NetworkTab.tsx:9]: The component name and filename should be the same.

File: src/containers/Network/NetworkTab.tsx (hunk 9 above)
Comment (mvadari) [reply src/containers/Network/NetworkTab.tsx:1]: The refactors can still be in a separate PR, though.

Comment (ckniffen) [reply src/containers/Network/NetworkTab.tsx:1]: Fair.

File: package.json [hunk 11]
```diff
@@ -5,13 +5,15 @@
   "dependencies": {
     "@fontsource/roboto": "^4.5.7",
     "@google-cloud/bigquery": "^5.5.0",
+    "@types/recharts": "^1.8.23",
```
Comment (mvadari) [inline package.json:8]: Types should be only dev dependencies

Comment (ckniffen) [inline package.json:8]: Types should be in dev dependencies.

File: public/locales/en-US/translations.json [hunk 12]
```diff
@@ -29,6 +29,7 @@
   "transaction_short": "TX",
   "nodes": "Nodes",
   "validators": "Validators",
+  "upgrade-status": "Upgrade Status",
```
Comment (ckniffen) [inline public/locales/en-US/translations.json:32]: Should have an underscore.  `Tabs.tsx` can be updated with `{t(title.replace('-', '_'))}` like in #270.

Comment (pdp2121) [reply public/locales/en-US/translations.json:32]: I will just replace - with _ for now. Tabs can be updated once #270 get merged

File: src/containers/Network/BarChartVersion.tsx [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+  }
+  return null
+}
+
+const BarChartVersion = (props: Props) => {
+  const { data } = props
+  const { t } = useTranslation()
+  const [posData, setposData] = useState<BarCoordinates>({ x: 0, y: 0 })
+  const grey = '#9BA2B0'
+  const purple = '#8884d8'
```
Comment (JST5000) [inline src/containers/Network/BarChartVersion.tsx:52]: Should these colors be defined in the css as well?

File: src/containers/Network/UpgradeStatus.tsx [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+  const aggregateData = (validators: any[]) => {
+    if (!validators) {
+      return []
+    }
+    let total = 0
+    const tempData: any[] = []
+    validators.reduce((aggregate, current) => {
+      if (!aggregate[current.server_version]) {
+        // eslint-disable-next-line no-param-reassign
```
Comment (JST5000) [inline src/containers/Network/UpgradeStatus.tsx:27]: For these eslint comments can you make sure to add comments explaining why it's necessary? (Or find a way to avoid them?)

I often find when I try explaining why I need to do them, I end up finding a way to avoid it :P

File: src/containers/Network/UpgradeStatus.tsx [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+          resp.data.filter((validator: any) => Boolean(validator.unl)).length,
+        )
+      })
+      .catch((e) => Log.error(e))
+  }
+
+  const updateValidators = (newValidations: any[]) => {
+    // @ts-ignore
+    setValidations(newValidations)
+    setVList((value: any) => {
```
Comment (JST5000) [inline src/containers/Network/UpgradeStatus.tsx:79]: What kind of value is this?

File: src/containers/Network/UpgradeStatus.tsx [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+        // @ts-ignore
+        <Hexagons data={validations} list={vList} />
+      )}
+
+      <div className="stat">
+        {validatorCount && (
+          <>
+            <span>{t('validators_found')}: </span>
+            <span>
+              {localizeNumber(validatorCount, language)}
```
Comment (JST5000) [inline src/containers/Network/UpgradeStatus.tsx:107]: Should this be used on all numbers (including the `unlCount` on line 111)?

Comment (ckniffen) [reply src/containers/Network/UpgradeStatus.tsx:107]: The difference in the usage might have been the size of the numbers.  Like unl count will likely never be too big.  I dont think we are consistent with localizeNumber.  Probably should be audited.

**Comment by intelliot** [comment]
Does this only chart the versions of the (~139) validators?

Can people also view a chart of the versions of all nodes on the network (~757)?

**Review by ckniffen — APPROVED** [review]

File: src/containers/Network/Validators.tsx [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+  useEffect(() => {
+    fetchData()
+    const interval = setInterval(fetchData, FETCH_INTERVAL_MILLIS)
+    return () => {
+      clearInterval(interval)
+    }
+  }, [])
+
+  const fetchData = () => {
+    const url = '/api/v1/validators?verbose=true'
```
Comment (jonathanlei) [inline src/containers/Network/Validators.tsx:29]: nit: this`fetchData` function is used also in `UpgradeStatus.tsx`, can it be extracted and passed in as a pop in `index.js`?

File: src/containers/Network/UpgradeStatus.tsx [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+      return aggregation
+    }, {})
+
+    return tempData
+      .map((item) => ({
+        label: item.server_version ? item.server_version : ' N/A ',
+        value: (item.count * 100) / total,
+        count: item.count,
+      }))
+      .sort((a, b) => (a.label > b.label ? 1 : -1))
```
[the remark this answers is not in the document]
Comment (pdp2121) [reply src/containers/Network/UpgradeStatus.tsx:45]: I added a space so that ' N/A ' will always appear left most of the x-axis (valid version should always start with a number).
There can be multiple 'N/A' server as version return null when a server is updating to a new version, based on my observation.

File: src/containers/Network/Validators.tsx (hunk 17 above)
Comment (ckniffen) [reply src/containers/Network/Validators.tsx:29]: That is true.  We eliminated quite a bit of duplication. @pdp2121 we should target that when we add a chart for nodes.

**Review by JST5000 — APPROVED** [review]
LGTM
